File: src/application.cpp
#include "dy4.h"
#include "application.h"

ApplicationLayer::ApplicationLater (){
    program_id.clear();
    program_type = "Undefined";
    program_service_group = "not0A";
    program_service_name_addresses = -1;
    program_service_names.clear();
}

void ApplicationLayer::get_message(char block_type,
                                   std::vector<bool> &block_data){
    // Process message based on the RDS block type and data within
    // flip data around
    std::reverse(block_data.begin(), block_data.end());
    if (block_type == "A"){
        extract_program_id(block_data);
    } else if (block_type == "B"){
        extract_program_type(std::copy(block_data.begin()+5, block_data.begin() + 9));
        extract_program_service_group(block_data.begin()+11, block_data.end());
        extract_program_service_name_address(block_data.begin(),block_data.begin()+1);
    } else if (block_type == "D"){
        extract_program_service_name(block_data);
    }
}

void list_to_int(std::vector<bool> &block_data,
                 int data){
    data = 0;
    for (unsigned int i=block_data.size(); i>0; i--){
        data *= 2;
        data += (int)(block_data[i]);
    }
}

void extract_program_id(std::vector<bool> &block_data){
    // Extracts the program identification code from block
    // convert block data into character
    int data;
    list_to_int(block_data, data);

    if ((program_id.size() == 0) && (data != 12)){
        return;
    } else if ((program_id.size() == 0) && (data == 12)){
        program_id.push_back(std::hex << data);
    } else if ((program_id.size() > 1) && (program.size() < 4)){
        program_id.push_back(std::hex << data);
    } else {
        return;
    }
}

void extract_program_type(std::vector<bool> &block_data){
    // Extracts the program type from block
    int data;
    list_to_int(block_data, data);
    int data;
    if (data == 1){
        program_type = "News";
    } else if (data == 2){
        program_type = "Information";
    } else if (data == 3){
        program_type = "Sport";
    } else if (data == 4){
        program_type = "Talk";
    } else if (data == 5){
        program_type = "Rock";
    } else if (data == 6){
        program_type = "Classic rock";
    } else if (data == 7){
        program_type = "Adult Hits";
    } else if (data == 8){
        program_type = "Soft Rock";
    } else if (data == 9){
        program_type = "Top 40";
    } else if (data == 10){
        program_type = "Country Music";
    } else if (data == 11){
        program_type = "Oldies Music";
    } else if (data == 12){
        program_type = "Soft Music";
    } else if (data == 13){
        program_type = "Nostalgia";
    } else if (data == 14){
        program_type = "Jazz";
    } else if (data == 15){
        program_type = "Classical";
    } else if (data == 16){
        program_type = "Rhythm & Blues Music";
    } else if (data == 17){
        program_type = "Soft Rhythm & Blues Music";
    }
}



void extract_program_service_name_address (std::vector<bool> &block_data){
    list_to_int(block_data, program_service_name_addresses);
}

void extract_program_service_group (std::vector<bool> &block_data){
    // extract program service name from the block
    int data;
    list_to_int(block_data, data);

    if (data == 0){
        program_service_group = "A0";
    } else {
        program_service_group = "notA0";
    }
}

void extract_program_service_name (std::vector<bool> &block_data){
    // extract the data from offset D
    int data;
    list_to_int(block_data, data);

    if (program_service_group == "A0"){
        program_service_names[program_service_names.size()-1] += (char)(data);
        if (program_service_name_address == 0){
            program_service.push_back("");
        }
}

File: model/application.py

class ApplicationLayer(object):
    def __init__(self):
        self.program_id = []
        self.program_type = "Undefined"
        self.program_service_group = "not0A"
        self.program_service_name_addresses = -1
        self.program_service_names = [""]

    def get_message(self, block_type, block_data):
        """Process message based on the RDS block type and data within"""
        # flip data around
        block_data = block_data[::-1]
        if block_type == "A":
            self.extract_program_id(block_data)
        elif block_type == "B":
            self.extract_program_type(block_data[5:10])
            self.extract_program_service_group(block_data[11:])
            self.extract_program_service_name_address(block_data[:2])
        elif block_type == "D":
            self.extract_program_service_name(block_data)

    def print_attr(self):
        print("program_id: ", self.program_id)
        print("program type: ", self.program_type)
        print("program service group: ", self.program_service_group)
        print("program service name addresses: ", self.program_service_name_addresses)
        print("program service names: ", self.program_service_names)
        
    def list_to_int(self, block_data):
        data = 0
        for item in block_data[::-1]:
            data *= 2
            data += int(item)
        return data

    def extract_program_id(self, block_data):
        """Extracts the program identification code from block"""
        # convert block data into character
        data = self.list_to_int(block_data)

        # Equivalent of C
        if len(self.program_id) == 0 and data != 12:
            return
        elif len(self.program_id) == 0 and data == 12:
            self.program_id.append(hex(data))
        elif len(self.program_id) > 1 and len(self.program_id) < 4:
            self.program_id.append(hex(data))
        else:
            return
    
    def extract_program_type(self, block_data):
        """Extracts the program type from block"""
        data = self.list_to_int(block_data)

        if data == 1:
            self.program_type = "News"
        elif data == 2:
            self.program_type = "Information"
        elif data == 3:
            self.program_type = "Sport"
        elif data == 4:
            self.program_type = "Talk"
        elif data == 5:
            self.program_type = "Rock"
        elif data == 6:
            self.program_type = "Classic rock"
        elif data == 7:
            self.program_type = "Adult Hits"
        elif data == 8:
            self.program_type = "Soft Rock"
        elif data == 9:
            self.program_type = "Top 40"
        elif data == 10:
            self.program_type = "Country Music"
        elif data == 11:
            self.program_type = "Oldies Music"
        elif data == 12:
            self.program_type = "Soft Music"
        elif data == 13:
            self.program_type = "Nostalgia"
        elif data == 14:
            self.program_type = "Jazz"
        elif data == 15:
            self.program_type = "Classical"
        elif data == 16:
            self.program_type = "Rhythm & Blues Music"
        elif data == 17:
            self.program_type = "Soft Rhythm & Blues Music"

    def extract_program_service_name_address(self, block_data):
        self.program_service_name_addresses = self.list_to_int(block_data)
    
    def extract_program_service_group(self, block_data):
        """extract program service name from the block"""
        data = self.list_to_int(block_data)

        if data == 0:
            self.program_service_group = "A0"
        else:
            self.program_service_group = "notA0"
    
    def extract_program_service_name(self, block_data):
        """extract the data from offset D"""
        data = self.list_to_int(block_data)
        if self.program_service_group == "A0":
            self.program_service_names[-1] += chr(data)
            if self.program_service_name_addresses == 0:
                self.program_service_names.append("")



